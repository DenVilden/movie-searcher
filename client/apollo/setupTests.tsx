import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from '../pages/_app';

// eslint-disable-next-line jest/prefer-spy-on
HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks = [], addTypeName = false }: RenderApolloOptions = {},
) =>
  render(
    <MockedProvider addTypename={addTypeName} mocks={mocks}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>,
  );

export * from '@testing-library/react';
