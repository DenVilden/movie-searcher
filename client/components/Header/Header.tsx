import { AppBar, Toolbar } from '@material-ui/core';
import Link from 'next/link';
import styled from 'styled-components';
import Favorites from '../Favorites/Favorites';
import SearchBar from '../SearchBar/SearchBar';

const StyledToolbar = styled(Toolbar)`
  padding: 0;

  ${(props) => props.theme.breakpoints.up('sm')} {
    padding-left: 8px;
    padding-right: 8px;
  }
`;

const StyledImg = styled.img`
  cursor: pointer;
`;

const Header = () => (
  <AppBar position="static">
    <StyledToolbar>
      <Link href="/">
        <StyledImg alt="logo" src="/logo.svg" />
      </Link>
      <SearchBar />
      <Favorites />
    </StyledToolbar>
  </AppBar>
);

export default Header;
