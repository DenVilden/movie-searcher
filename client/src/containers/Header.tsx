import React from 'react';
import { AppBar, Toolbar } from '@material-ui/core';
import { useSetInputValueMutation } from '../generated/mutations.generated';
import { useGetInputValueQuery } from '../generated/queries.generated';
import Favorites from './Favorites';
import Logo from '../components/Logo/Logo';
import SearchBar from '../components/SearchBar/SearchBar';
import SearchResults from './SearchResults';

const Header = () => {
  const { data } = useGetInputValueQuery();
  const [setInputValue] = useSetInputValueMutation();

  return data ? (
    <>
      <AppBar position="static">
        <Toolbar>
          <Logo />
          <SearchBar
            inputValue={data.inputValue}
            setInputValue={value => setInputValue({ variables: { value } })}
          />
          <Favorites />
        </Toolbar>
      </AppBar>
      {data.inputValue && <SearchResults query={data.inputValue} />}
    </>
  ) : null;
};

export default Header;
