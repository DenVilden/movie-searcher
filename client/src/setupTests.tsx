import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from './pages/_app';
import defaultResolvers from './graphql/resolvers';

window.HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
  cache?: any;
  resolvers?: any;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks, addTypeName, resolvers, cache }: RenderApolloOptions = {}
) =>
  render(
    <MockedProvider
      addTypename={addTypeName}
      cache={cache}
      mocks={mocks}
      resolvers={resolvers || defaultResolvers}
    >
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>
  );

export * from '@testing-library/react';
