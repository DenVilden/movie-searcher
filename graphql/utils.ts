export function paginateResults(data: any, size = 4, cursor: number | null) {
  let results = [];

  if (!cursor) {
    results = data.results.slice(0, size);
  } else {
    results = data.results.slice(0, cursor + size);
  }

  return {
    ...data,
    cursor: results.length,
    hasMore: results.length !== data.results.length,
    results,
  };
}

export function attachPoster(path: string, size = 200) {
  return path ? `https://image.tmdb.org/t/p/w${size}${path}` : null;
}
