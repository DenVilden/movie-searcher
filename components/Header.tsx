import { useReactiveVar } from '@apollo/client'
import { css } from '@emotion/react'
import { AppBar, IconButton, Toolbar } from '@material-ui/core'
import {
  Brightness4 as DarkIcon,
  Brightness7 as LightIcon,
} from '@material-ui/icons'
import Link from 'next/link'

import { prefersDarkModeVar } from '@/apollo/client'

import Favorites from './Favorites'
import Logo from './Logo'
import SearchBar from './SearchBar'

export default function Header() {
  const prefersDarkMode = useReactiveVar(prefersDarkModeVar)

  const toggleTheme = () => {
    const toggle = !prefersDarkMode
    prefersDarkModeVar(toggle)
    localStorage.setItem('darkMode', JSON.stringify(toggle))
  }

  return (
    <AppBar color="inherit" position="fixed">
      <Toolbar
        css={css`
          padding: 0 4px;
        `}
      >
        <Link href="/">
          <IconButton aria-label="go to home page">
            <Logo prefersDarkMode={prefersDarkMode} />
          </IconButton>
        </Link>
        <SearchBar />
        <IconButton
          aria-label="theme switch"
          css={css`
            margin-left: auto;
          `}
          onClick={toggleTheme}
        >
          {prefersDarkMode ? <LightIcon /> : <DarkIcon />}
        </IconButton>
        <Favorites />
      </Toolbar>
    </AppBar>
  )
}
