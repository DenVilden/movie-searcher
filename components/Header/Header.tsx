import { AppBar, Toolbar, IconButton } from '@material-ui/core';
import Link from 'next/link';
import styled from '@emotion/styled';
import {
  MovieFilter as MovieIcon,
  Brightness7 as LightIcon,
  Brightness4 as DarkIcon,
} from '@material-ui/icons';
import { memo } from 'react';
import { useReactiveVar } from '@apollo/client';
import { Favorites, SearchBar } from '..';
import { prefersDarkModeVar } from '../../apollo';

const StyledToolbar = styled(Toolbar)`
  padding: 0;

  ${(props) => props.theme.breakpoints.up('sm')} {
    padding-left: 8px;
    padding-right: 8px;
  }
`;

const StyledMovieIcon = styled(MovieIcon)`
  cursor: pointer;
  font-size: 60px;
`;

const StyledIconButton = styled(IconButton)`
  margin-left: auto;
`;

function Header() {
  const prefersDarkMode = useReactiveVar(prefersDarkModeVar);

  const toggleTheme = () => {
    const toggle = !prefersDarkMode;
    prefersDarkModeVar(toggle);
    localStorage.setItem('darkMode', JSON.stringify(toggle));
  };

  return (
    <AppBar position="static" color={prefersDarkMode ? 'inherit' : 'primary'}>
      <StyledToolbar>
        <Link href="/">
          <StyledMovieIcon />
        </Link>
        <SearchBar />
        <StyledIconButton aria-label="theme switch" onClick={toggleTheme}>
          {prefersDarkMode ? (
            <DarkIcon />
          ) : (
            <LightIcon style={{ color: 'white' }} />
          )}
        </StyledIconButton>
        <Favorites />
      </StyledToolbar>
    </AppBar>
  );
}

export default memo(Header);
