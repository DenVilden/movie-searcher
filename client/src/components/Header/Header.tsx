/* eslint-disable react/jsx-props-no-spreading */
import { AppBar, Toolbar, TextField, Button } from "@material-ui/core";
import { useState } from "react";
import {
  AutocompleteInputChangeReason,
  AutocompleteRenderInputParams,
  Autocomplete,
  AutocompleteChangeReason,
} from "@material-ui/lab";
import { useRouter } from "next/router";
import Link from "next/link";
import styled from "styled-components";
import { fade } from "@material-ui/core/styles";
import { useReactiveVar } from "@apollo/client";
import Favorites from "../Favorites/Favorites";
import { useGetMoviesSearchLazyQuery } from "../../graphql";
import { autocompleteVar } from "../../apollo";

const StyledToolbar = styled(Toolbar)`
  padding: 0;

  ${(props) => props.theme.breakpoints.up("sm")} {
    padding-left: 8px;
    padding-right: 8px;
  }
`;

const StyledAutocomplete = styled((props) => <Autocomplete {...props} />)`
  width: 100%;

  ${(props) => props.theme.breakpoints.up("md")} {
    width: 70%;
  }

  input {
    color: #fafafa;
  }
`;

const StyledInputBase = styled(TextField)`
  background-color: ${(props) => fade(props.theme.palette.common.white, 0.15)};
  border-radius: ${(props) => props.theme.shape.borderRadius}px;

  ${(props) => props.theme.breakpoints.up("sm")} {
    margin-left: ${(props) => props.theme.spacing(2)}px;
  }

  :hover {
    background-color: ${(props) =>
      fade(props.theme.palette.common.white, 0.25)};
  }
`;

export const Header = () => {
  const [inputValue, setInputValue] = useState("");
  const autocompleteValue = useReactiveVar(autocompleteVar);

  const [fetchMovies, { data, loading, error }] = useGetMoviesSearchLazyQuery();

  const router = useRouter();

  return (
    <AppBar position="static">
      <StyledToolbar>
        <Link href="/">
          <Button title="logo" onClick={() => autocompleteVar("")}>
            <img alt="logo" src="/logo.svg" />
          </Button>
        </Link>
        <StyledAutocomplete
          autoHighlight
          blurOnSelect
          freeSolo
          value={autocompleteValue}
          loading={loading}
          onChange={(
            _evt: React.ChangeEvent<HTMLLIElement>,
            value: string,
            reason: AutocompleteChangeReason
          ) => {
            if (reason === "select-option") {
              const id = data?.moviesSearch.results.find(
                (movie) => movie.title === value
              )?.id;
              autocompleteVar(value);
              router.push(`/movie/${id}`);
            }
          }}
          onInputChange={(
            _evt: React.ChangeEvent<HTMLInputElement>,
            value: string,
            reason: AutocompleteInputChangeReason
          ) => {
            if (reason === "input" && value) {
              setInputValue(value);
              fetchMovies({ variables: { query: value, pageSize: 8 } });
            } else {
              setInputValue("");
            }
          }}
          open={!!inputValue}
          options={data?.moviesSearch.results.map((movie) => movie.title) || []}
          renderInput={(params: AutocompleteRenderInputParams) => (
            <StyledInputBase
              {...params}
              error={!!error}
              variant="outlined"
              margin="dense"
              value={inputValue}
              placeholder="type a movie name..."
              fullWidth
            />
          )}
        />
        <Favorites />
      </StyledToolbar>
    </AppBar>
  );
};

const WithHeader = (Component: React.ComponentType) => ({ ...pageProps }) => (
  <>
    <Header />
    <Component {...pageProps} />
  </>
);

export default WithHeader;
