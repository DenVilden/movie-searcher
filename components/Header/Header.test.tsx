import Header from './Header';
import { renderApollo } from '../../utils/setupTests';

const mock = {
  id: 1,
  backdrop_path: null,
  poster_path: null,
  title: 'test',
  overview: 'test data',
  budget: '0',
  revenue: '0',
  vote_average: 5,
  release_date: '2020',
  similar: {
    results: [
      {
        id: 1,
        title: 'test',
        release_date: '2020',
        poster_path: null,
      },
    ],
  },
};

describe('header', () => {
  it('should take a snapshot', () => {
    const { asFragment } = renderApollo(<Header />);

    const element = asFragment();

    expect(element).toMatchSnapshot();
  });

  it('should set initial favorites', () => {
    localStorage.setItem('favorites', JSON.stringify([mock]));

    const { getByText } = renderApollo(<Header />);

    expect(getByText('1')).toBeTruthy();
  });
});
