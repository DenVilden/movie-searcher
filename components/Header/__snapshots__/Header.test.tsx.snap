// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header should take a snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1y9n3ry-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-tao4pc-MuiToolbar-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1p9831u-MuiSvgIcon-root"
        data-testid="MovieFilterIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 4l2 3h-3l-2-3h-2l2 3h-3l-2-3H8l2 3H7L5 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4h-4zm-6.75 11.25L10 18l-1.25-2.75L6 14l2.75-1.25L10 10l1.25 2.75L14 14l-2.75 1.25zm5.69-3.31L16 14l-.94-2.06L13 11l2.06-.94L16 8l.94 2.06L19 11l-2.06.94z"
        />
      </svg>
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root css-pl0oq3-MuiAutocomplete-root"
        role="combobox"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bpzt7r-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <div
          class="MuiFormControl-root MuiFormControl-marginNone MuiFormControl-fullWidth MuiTextField-root css-2ztzun-MuiFormControl-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-4a33cb-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused MuiAutocomplete-input MuiAutocomplete-inputFocused css-b55ri7-MuiInputBase-input-MuiOutlinedInput-input"
              id="autocomplete"
              placeholder="Search..."
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1o2jng6-MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="css-1br1jja"
                style="width: 0.01px;"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        aria-label="theme switch"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-14dhwkx-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="Brightness7Icon"
            focusable="false"
            style="color: white;"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="open favorites"
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1drqkcs-MuiButtonBase-root-MuiIconButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
        >
          <span
            class="MuiBadge-root css-1c32n2y-MuiBadge-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
            <span
              class="MuiBadge-colorSecondary MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRightRectangular MuiBadge-invisible css-13j2lht-MuiBadge-badge"
            />
          </span>
        </span>
      </button>
    </div>
  </header>
</DocumentFragment>
`;
